File: src/contests/cg/puzzle/swift/fibonaccis-rabbit.swift
// 🎮 CodinGame Puzzle - fibonaccis-rabbit
// https://www.codingame.com/training/easy/fibonaccis-rabbit

let l1 = readLine()!.split(separator: " ").map { Int($0)! }
let f0 = l1[0], n = l1[1]
let l2 = readLine()!.split(separator: " ").map { Int($0)! }
let a = l2[0], b = l2[1]
// FN can exceed 2^63 (but is < 2^64), so accumulate in UInt64.
var f = [UInt64](repeating: 0, count: n + 1)
f[0] = UInt64(f0)
for i in stride(from: 1, through: n, by: 1) {
    var total: UInt64 = 0
    for k in a...b {
        if i - k >= 0 {
            total += f[i - k]
        }
    }
    f[i] = total
}
print(f[n])

File: src/contests/cg/puzzle/go/fibonaccis-rabbit.go
// 🎮 CodinGame Puzzle - fibonaccis-rabbit
// https://www.codingame.com/training/easy/fibonaccis-rabbit

package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	var f0, n int
	fmt.Fscan(reader, &f0, &n)
	var a, b int
	fmt.Fscan(reader, &a, &b)
	// FN can exceed 2^63 (but is < 2^64), so accumulate in uint64.
	f := make([]uint64, n+1)
	f[0] = uint64(f0)
	for i := 1; i <= n; i++ {
		var total uint64 = 0
		for k := a; k <= b; k++ {
			if i-k >= 0 {
				total += f[i-k]
			}
		}
		f[i] = total
	}
	fmt.Println(f[n])
}
